content/posts/2020/02/thai-in-2562/index.eng.md

```diff
@@ -0,0 +1,94 @@
+
+{{<
+figure
+  src="1.png"
+  caption="Screenshot of Our Participant"
+>}}
+
+We checked the screenshot our user sent us and realized one strange thing. The date was 2562 instead of 2019. So we came up with a hypothesis. For some reason, the file system was broken and the hash of the block header was being miscalculated.
+
+To verify this, we asked for permission to investigate the problem directly through remote desktop.
+
+
+## The Date Was Not Wrong
+
+Fortunately, the user experiencing this problem kindly accepted our request for remote access. And once we were able to access the desktop, we saw that the date was still in 2562 as it appeared in the screenshot. First, we opened the Control Panel to sync time. Like most modern operating systems, Windows has the ability to synchronize time over the network. But when we resynchronized time, the control panel and the system's year didn't change and remained in 2562. Notably, apart from the year, date and time were no different from the system on our side.
+
+We decided to explore the problem a little bit more, and while checking out the control panel, one detail caught our attention.
+
+{{<
+figure
+  src="2.png"
+  caption="Year Displayed as 2019 in Gregorian calendar"
+>}}
+
+No one in our development team could read Thai, but looking at "Date in Gregorian”, we noticed that "2562" was a different expression of "2019." So we changed the format, and as expected, December 16th, 2019 was displayed. And when we launched the game for testing, it went smoothly at the IBD stage.
+
+## Buddhist Calendar
+
+Now that we had found a clue and a way to reproduce the issue, we suggested our user to set up the region in the US for the time being, and thankfully, our user agreed to our proposal.
+
+To reproduce this problem on our side, we changed the regional setting of our OS to Thailand in the local development environment and then ran the unit test of Libplanet. Sure enough, we were able to see some of them fail. The most crucial problem was that the hash of the same block was changing. As we looked closer, the result of [serializing `Block<T>.Timestamp` field][1] in the process of creating a hash input was different than expected. The behavior of the `DateTimeOffset.ToString()` method was affected by the locale of the operating system.
+
+In Indochina Peninsula, Buddhism takes the same place as Christianity in Europe. So, instead of the [Gregorian calendar], which uses the birth of Jesus Christ as an epoch, countries in Indochina Peninsula had used the [Buddhist calendar] which takes Buddha’s attainment of Parinirvana(nirvana after death) as an epoch. Although countries like Cambodia and Laos eventually took on the Gregorian calendar, Thailand still uses the [Thai solar calendar], which is a solar modification of the traditional Buddhist lunar calendar. Buddha’s Parinirvana was in 543 BC, and so the year 2019 A.D. becomes 2562 under the Thai solar calendar.
```

Comment:
```suggestion
In Indochina Peninsula, Buddhism takes the same place as Christianity in Europe. So, instead of the [Gregorian calendar], which uses the birth of Jesus Christ as an epoch, countries in Indochina Peninsula had used the [Buddhist calendar] which takes Buddha’s attainment of parinirvana (nirvana after death) as an epoch. Although countries like Cambodia and Laos eventually took on the Gregorian calendar, Thailand still uses the [Thai solar calendar], which is a solar modification of the traditional Buddhist lunar calendar. Buddha’s parinirvana was in 543 BC, and so the year 2019 A.D. becomes 2562 under the Thai solar calendar.
```